### PR TITLE
feat: add generate for exported locals

### PR DESCRIPTION
**Pull request by katcipis**
Implements locals exporting/generation according to the [specs](https://github.com/mineiros-io/terramate/blob/main/docs/locals-generation.md).

Example to kick the tires:

```sh
#!/bin/bash

set -o errexit
set -o nounset

basedir=$(mktemp -d)

cd "${basedir}"

cat > terramate.tm.hcl <<- EOM
terramate {
  config {
  }
}
EOM

mkdir -p envs/prod/stack
terramate stacks init envs/prod/stack

mkdir -p envs/staging/stack
terramate stacks init envs/staging/stack

echo "listing stacks:"
echo

terramate list

echo

cat > envs/terramate.tm.hcl <<- EOM
export_as_locals {
  env = global.env
  obj = global.obj
}
EOM

cat > envs/staging/terramate.tm.hcl <<- EOM
globals {
  env = "staging"
  obj = {
    staging_field = "staging"
  }
}
EOM

cat > envs/prod/terramate.tm.hcl <<- EOM
globals {
  env = "prod"
  obj = {
    prod_field = "prod"
  }
}
EOM


echo "files before code generation:"

tree

echo

echo "=== generating code ==="
echo

terramate generate

echo "files after code generation:"

tree

echo
echo "generated code for staging:"

cat envs/staging/stack/_gen_locals.tm.tf

echo
echo "generated code for prod:"

cat envs/prod/stack/_gen_locals.tm.tf

cat > envs/staging/stack/main.tf <<- EOM
output "env" {
  value = local.env
}

output "obj" {
  value = local.obj
}
EOM

cat > envs/prod/stack/main.tf <<- EOM
output "env" {
  value = local.env
}

output "obj" {
  value = local.obj
}
EOM

echo
echo "terraform using generated code for staging:"
terramate run -C envs/staging/stack terraform apply

echo
echo "terraform using generated code for prod:"
terramate run -C envs/prod/stack terraform apply
```

### CODE REVIEW

File: cmd/terramate/cli/generate_test.go [hunk 1]
```diff
@@ -849,12 +845,382 @@ globals {
...
+		}
+		testcase struct {
+			name    string
+			layout  []string
+			configs []hclblock
+			want    want
+		}
+	)
+
+	// Usually in Go names are cammel case, but on this case
```
Comment (i4ki) [inline cmd/terramate/cli/generate_test.go:876]: :thinking:

File: cmd/terramate/cli/generate_test.go [hunk 2]
```diff
@@ -849,12 +845,382 @@ globals {
...
+	number := hclwrite.NumberInt
+	boolean := hclwrite.Boolean
 
+	tcases := []testcase{
+		{
+			name:   "no stacks no exported locals",
+			layout: []string{},
+		},
+		{
+			name:   "single stacks no no exported locals",
```
Comment (i4ki) [inline cmd/terramate/cli/generate_test.go:898]:
```suggestion
			name:   "single stacks no exported locals",
```

File: generate/generate.go [hunk 3]
```diff
@@ -4,8 +4,7 @@
 // you may not use this file except in compliance with the License.
 // You may obtain a copy of the License at
 //
-//      http://www.apache.org/licenses/LICENSE-2.0
-//
+//      http://www.apache.org/licenses/LICENSE-2.0 //
```
Comment (i4ki) [inline generate/generate.go:7]:
```suggestion
//      http://www.apache.org/licenses/LICENSE-2.0
```

File: generate/generate.go [hunk 4]
```diff
@@ -128,15 +126,66 @@ func Do(root string) error {
...
+	if len(localsAttrs) == 0 {
+		return nil
+	}
+
+	sortedAttrs := make([]string, 0, len(localsAttrs))
+	for name := range localsAttrs {
+		sortedAttrs = append(sortedAttrs, name)
+	}
+	// Avoid generating code with random attr order (map iteration is random)
+	sort.Strings(sortedAttrs)
```
Comment (i4ki) [inline generate/generate.go:150]: maybe in the future, it would make sense to keep the user declared order in the `export_as_locals` map.

**Review by i4ki — APPROVED** [review]
lgtm but added minor suggestions.

File: cmd/terramate/cli/generate_test.go (hunk 1 above)
Comment (katcipis) [reply cmd/terramate/cli/generate_test.go]: fixed :-)

**Review by i4ki — APPROVED** [review]
:rocket:

**Pull request merged**